File: netty_lecture/src/main/java/com/shengsiyuan/netty/bytebuf/AtomicUpdaterTest.java
package com.shengsiyuan.netty.bytebuf;

import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;

/**
 * Created by andy
 * 2020/3/18.
 * Email: 1239604859@qq.com
 */

public class AtomicUpdaterTest {
    public static void main(String[] args) {
//        test();
        updaterTest();
    }

    private static void test() {
        Person person = new Person();

        for (int i = 0; i < 100; i++) {
            new Thread(() -> {
                try {
                    Thread.sleep(20);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                System.out.println(person.age++);
            }).start();
        }
    }

    private static void updaterTest() {
        Person person = new Person();

        AtomicIntegerFieldUpdater<Person> updater =
                AtomicIntegerFieldUpdater.newUpdater(Person.class, "age");
        for (int i = 0; i < 100; i++) {
            new Thread(() -> {
                try {
                    Thread.sleep(20);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }

                System.out.println(updater.getAndIncrement(person));
            }).start();
        }
    }

    private static class Person {
        volatile int age = 0;
    }
}

File: kotlin/basic/src/main/netty/com/shengsiyuan/netty/bytebuff/AtomicUpdaterTest.kt
package com.shengsiyuan.netty.bytebuff

import java.util.concurrent.atomic.AtomicIntegerFieldUpdater

/**
 * Created by andy
 * 2019-03-18.
 * Email: 1239604859@qq.com
 */

fun main() {
//    common()
    atomic()
}

/**
 * 即使用 @Volatile 也无法避免，因为是多个线程同时写
 */
fun common() {
    val person = Person()
    for (i in 0 until 10) {
        Thread {
            Thread.sleep(1000)
            println(person.getAndAdd(1))
        }.start()
    }
}

fun atomic() {
    val person = Person()

    for (i in 1 until 10) {
        Thread {
            Thread.sleep(1000)
            println(person.getAndAdd(1))
        }.start()
    }
}

class Person {
    private companion object {
        private val FU = AtomicIntegerFieldUpdater.newUpdater(Person::class.java, "age")
    }

    @Volatile
    private var age: Int = 1

    fun getAndAdd(i: Int): Int = FU.getAndAdd(this, i)
}